bee-network/src/swarm/protocols/iota_gossip/handler.rs

```diff
@@ -0,0 +1,205 @@
+        debug!("gossip handler: responding to listen protocol request.");
+
+        SubstreamProtocol::new(IotaGossipProtocolUpgrade::new(self.info.clone()), ())
+    }
+
+    /// **libp2p docs**:
+    ///
+    /// Injects an event coming from the outside in the handler.
+    fn inject_event(&mut self, incoming_event: IotaGossipHandlerInEvent) {
+        debug!("gossip handler: received in-event: {:?}", incoming_event);
+
+        let IotaGossipHandlerInEvent { origin } = incoming_event;
+
+        // We only send the upgrade request if this handler belongs to an outbound connection.
+        if origin == Origin::Outbound {
+            let send_request = ProtocolsHandlerEvent::OutboundSubstreamRequest {
+                protocol: SubstreamProtocol::new(IotaGossipProtocolUpgrade::new(self.info.clone()), ()),
+            };
+
+            debug!("gossip handler: sending protocol upgrade request.");
+
+            self.events.push_back(send_request);
+        }
+    }
+
+    /// **libp2p docs**:
+    ///
+    /// Injects the output of a successful upgrade on a new inbound substream.
+    fn inject_fully_negotiated_inbound(&mut self, new_inbound: NegotiatedSubstream, _: Self::InboundOpenInfo) {
+        let negotiated_inbound = ProtocolsHandlerEvent::Custom(IotaGossipHandlerEvent::UpgradeCompleted {
+            substream: Box::new(new_inbound),
+        });
+
+        debug!("gossip handler: fully negotiated inbound.");
+
+        self.events.push_back(negotiated_inbound);
+    }
+
+    /// **libp2p docs**:
+    ///
+    /// Injects the output of a successful upgrade on a new outbound substream.
+    ///
+    /// The second argument is the information that was previously passed to
+    /// [`ProtocolsHandlerEvent::OutboundSubstreamRequest`].
+    fn inject_fully_negotiated_outbound(&mut self, new_outbound: NegotiatedSubstream, _: Self::OutboundOpenInfo) {
+        let negotiated_outbound = ProtocolsHandlerEvent::Custom(IotaGossipHandlerEvent::UpgradeCompleted {
+            substream: Box::new(new_outbound),
+        });
+
+        debug!("gossip handler: fully negotiated outbound.");
+
+        self.events.push_back(negotiated_outbound);
+    }
+
+    /// **libp2p docs**:
+    ///
+    /// Notifies the handler of a change in the address of the remote.
+    fn inject_address_change(&mut self, new_address: &Multiaddr) {
+        debug!("gossip handler: new address: {}", new_address);
+    }
+
+    /// **libp2p docs**:
+    ///
+    /// Indicates to the handler that upgrading an outbound substream to the given protocol has failed.
+    fn inject_dial_upgrade_error(
+        &mut self,
+        _: Self::OutboundOpenInfo,
+        e: ProtocolsHandlerUpgrErr<<Self::OutboundProtocol as OutboundUpgrade<NegotiatedSubstream>>::Error>,
+    ) {
+        debug!("gossip handler: outbound upgrade error: {:?}", e);
+
+        // self.events.push_back(ProtocolsHandlerEvent::Close(e));
```

Comment:
How come this is commented out – is it okay to just leave this error as a `debug` log? If so, is it worth removing and detailing why this is fine?

Reply:
I added a `TODO`, because I urgently need to work on something else. I added a `TODO` to finsish this with the next patch version. I tried implementing it....then realized it needs more careful thinking, but didn't wanna lose the work I already spent on it.